common/dpcpp/slab_hash.hpp

```diff
@@ -0,0 +1,182 @@
+#pragma once
+
+#include <CL/sycl.hpp>
+#include <algorithm>
+#include "dpcpp_common.hpp"
+
+#define SUBGROUP_SIZE 16
+#define CONST 64
+#define SLAB_SIZE CONST * SUBGROUP_SIZE
+
+#define CLUSTER_SIZE 1024
+
+#define BUCKETS_COUNT 128
+
+#define EMPTY_UINT32_T 4294967295
+
+using std::pair;
+
+template <size_t A, size_t B, size_t P>
+struct Hasher {
+    size_t operator()(const uint32_t &k) { return ((A * k + B) % P) % BUCKETS_COUNT; };
+};
+
+template <typename T>
+struct SlabNode {
+    SlabNode() = default;            
+    SlabNode(T el) {
+        for (int i = 0; i < SLAB_SIZE; i++) {
+            data[i] = el;
+        }
+    }
+
+    T data[SLAB_SIZE];
+    sycl::global_ptr<SlabNode<T>> next = nullptr;
+};
+
+template <typename T>
+struct SlabList {
+    SlabList() = default;
+
+    sycl::global_ptr<SlabNode<T>> root;
+};
+
+template <typename K, typename T, typename Hash>
+class SlabHash {
+public:
+    SlabHash() = default;
+    SlabHash(K empty, Hash hasher, 
+            sycl::global_ptr<SlabList<pair<K, T>>> lists,
+            sycl::nd_item<1> & it, 
```

Comment:
Does this approach work across kernels? Thinking of building and then probing in a separate kernel.

Reply:
I'm building and probing in separate kernels in [tests](https://github.com/bagrorg/dwarf_bench/blob/cd2776ca0d262de4cc8fa4a7b6623ad6527340c2/tests/slab_tests.cpp#L150) and [build](https://github.com/bagrorg/dwarf_bench/blob/cd2776ca0d262de4cc8fa4a7b6623ad6527340c2/hash/slab_hash_build.cpp#L51)
Or did I misunderstand the question? :)

Reply:
No, this just looks weird to me. What happens if those kernels are run with different iteration space?

Reply:
Hm.. inside the kernel, I use `nd_item` only to get the local index. then it is used exclusively for splitting work during node processing.
In principle, `nd_item` is not needed as a field.
The differences between two separate kernels will only be in global indexes, but I don't use them here. Or what do you mean by iteration space?